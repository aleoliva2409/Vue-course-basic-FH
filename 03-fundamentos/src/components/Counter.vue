<template>
  <h2>{{ customTitle }}</h2>
  <p>{{ counter }} <sup>2</sup> = {{ counter * counter }}</p>
  <p>{{ counter }} <sup>2</sup> = {{ getSquareValue() }}</p>
  <p>{{ counter }} <sup>2</sup> = {{ squareCounter }}</p>
  <p data-testid="counter">{{ counter }}</p>
  <!-- <button v-on:click="decrease"> -1 </button> -->
  <!-- <button v-on:click="increase"> +1 </button> -->
  <button data-decrement="decrement" @click="decrease">-1</button> <!-- new form -->
  <button data-increment="increment" @click="increase">+1</button> <!-- new form -->
</template>

<script>
export default {
  // props: ['title', 'start'],
  props: {
    title: String,
    start: {
      type: Number,
      default: 50,
      // required: true,
      validator(value) {
        return value > 25;
      },
    },
  },
  name: 'Counter',
  data() {
    return {
      counter: this.start,
    };
  },
  methods: {
    getSquareValue() {
      return this.counter * this.counter;
    },
    decrease() {
      this.counter = this.counter - 1;
    },
    increase() {
      this.counter = this.counter + 1;
    },
  },
  computed: {
    squareCounter() {
      return this.counter * this.counter;
    },
    customTitle() {
      return this.title ? this.title : 'Counter';
    },
  },
};
</script>
