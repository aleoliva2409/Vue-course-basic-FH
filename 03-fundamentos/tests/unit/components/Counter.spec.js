import { shallowMount } from '@vue/test-utils';
import Counter from '@/components/Counter.vue';

describe('Counter component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Counter);
  });

  test('should match with snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('h2 should have default value', () => {
    expect(wrapper.find('h2').exists()).toBeTruthy();

    const h2Value = wrapper.find('h2').text();

    expect(h2Value).toBe('Counter');
  });

  test('default value should be 50 in p element', () => {
    // const elements = wrapper.findAll('p');
    const value = wrapper.find('[data-testid="counter"]').text();

    expect(value).toBe('50');
  });

  test('should increment and decrement counter value', async () => {
    const decrementBtn = wrapper.find('[data-decrement="decrement"]');
    const incrementBtn = wrapper.find('[data-increment="increment"]');

    await incrementBtn.trigger('click');

    let value = wrapper.find('[data-testid="counter"]').text();

    expect(value).toBe('51');

    await decrementBtn.trigger('click');
    await decrementBtn.trigger('click');

    // ? read value again because that varialbe is no reactive
    value = wrapper.find('[data-testid="counter"]').text();

    expect(value).toBe('49');
  });

  test('should set the default value ', () => {
    // ? read component props
    const { start } = wrapper.props();

    const value = wrapper.find('[data-testid="counter"]').text();

    expect(Number(value)).toBe(start);
  });

  test('should show title prop', () => {
    const title = 'Vamos los pibes';

    const wrapper = shallowMount(Counter, {
      props: {
        title,
      },
    });

    const titleValue = wrapper.find('h2').text();

    expect(titleValue).toBe(title);
  });
});
