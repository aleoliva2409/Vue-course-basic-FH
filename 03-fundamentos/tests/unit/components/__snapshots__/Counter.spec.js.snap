// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component should match with snapshot 1`] = `
<h2>Counter</h2>
<p>50 <sup>2</sup> = 2500</p>
<p>50 <sup>2</sup> = 2500</p>
<p>50 <sup>2</sup> = 2500</p>
<p data-testid="counter">50</p><!-- <button v-on:click="decrease"> -1 </button> -->
<!-- <button v-on:click="increase"> +1 </button> --><button data-decrement="decrement">-1</button><!-- new form --><button data-increment="increment">+1</button><!-- new form -->
`;
